crates/rpc/rpc/src/eth/api/mod.rs

```diff
@@ -438,4 +453,144 @@ struct EthApiInner<Provider, Pool, Network> {
+#[derive(Debug, Clone)]
+pub struct FeeHistoryCache<Provider> {
+    lower_bound: Arc<AtomicU64>,
+    upper_bound: Arc<AtomicU64>,
+    config: FeeHistoryCacheConfig,
+    provider: Provider,
+    entries: Arc<tokio::sync::RwLock<BTreeMap<u64, SealedHeader>>>,
+}
+
+impl<Provider> FeeHistoryCache<Provider>
+where
+    Provider: BlockReaderIdExt
+        + ChainSpecProvider
+        + StateProviderFactory
+        + EvmEnvProvider
+        + BlockNumReader
+        + 'static,
+{
+    /// Creates new FeeHistoryCache instance, initialize it with the mose recent data, set bounds
+    pub fn new(config: FeeHistoryCacheConfig, provider: Provider) -> Self {
+        let mut init_tree_map = BTreeMap::new();
+
+        let last_block_number = provider.last_block_number().unwrap_or(0);
+
+        let start_block = if last_block_number > config.max_blocks {
+            last_block_number - config.max_blocks
+        } else {
+            0
+        };
+
+        let headers =
+            provider.sealed_headers_range(start_block..=last_block_number).unwrap_or_default();
+
+        let (mut lower_bound, mut upper_bound) = (0, 0);
+        for header in headers {
+            init_tree_map.insert(header.number, header.clone());
+            upper_bound = header.number;
+            if lower_bound == 0 {
+                lower_bound = header.number;
+            }
+        }
+
+        let entries = Arc::new(tokio::sync::RwLock::new(init_tree_map));
+
+        let upper_bound = Arc::new(AtomicU64::new(upper_bound));
+        let lower_bound = Arc::new(AtomicU64::new(lower_bound));
+
+        FeeHistoryCache { config, entries, provider, upper_bound, lower_bound }
+    }
+
+    /// Processing of the arriving blocks
+    pub async fn on_new_block(&self, block: &SealedBlock) {
+        let mut entries = self.entries.write().await;
+        entries.insert(block.number, block.header.clone());
+        while entries.len() > self.config.max_blocks as usize {
+            entries.pop_first();
+        }
+        if entries.len() == 0 {
+            self.upper_bound.store(0, SeqCst);
+            self.lower_bound.store(0, SeqCst);
+            return
+        }
+        let upper_bound = *entries.last_entry().expect("Contains be at least one entry").key();
+        let lower_bound = *entries.first_entry().expect("Contains at least one entry").key();
+        self.upper_bound.store(upper_bound, SeqCst);
+        self.lower_bound.store(lower_bound, SeqCst);
+    }
+
+    /// Collect fee history for given range. It will try to use a cache to take the most recent
+    /// headers or if the range is out of caching config it will fallback to the database provider
+    pub async fn get_history(
+        &self,
+        start_block: u64,
+        end_block: u64,
+    ) -> RethResult<Vec<SealedHeader>> {
+        let headers: Vec<SealedHeader>;
+
+        let lower_bound = self.lower_bound.load(SeqCst);
+        let upper_bound = self.upper_bound.load(SeqCst);
```

Comment:
we can replace these with bounds check of the first/last entry for now,

let's move this to functions `fn {lower,upper}_bound` so we can reintroduce the Arc atomics later